File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/HistoryTableConnection.java
/*
 * Copyright (c) 2016. Eli Connelly
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.emogoth.android.phone.mimi.db;


import android.content.ContentValues;
import android.database.sqlite.SQLiteDatabase;
import android.text.TextUtils;
import android.util.Log;

import com.activeandroid.query.Delete;
import com.activeandroid.query.From;
import com.activeandroid.query.Select;
import com.emogoth.android.phone.mimi.app.MimiApplication;
import com.emogoth.android.phone.mimi.db.model.History;
import com.mimireader.chanlib.models.ChanPost;
import com.squareup.sqlbrite3.BriteDatabase;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.TimeUnit;

import io.reactivex.BackpressureStrategy;
import io.reactivex.Flowable;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Single;
import io.reactivex.functions.Function;

import static com.emogoth.android.phone.mimi.db.ActiveAndroidSqlBriteBridge.runQuery;


public class HistoryTableConnection {
    public static final String LOG_TAG = HistoryTableConnection.class.getSimpleName();

    public static final int HISTORY = 1;
    public static final int BOOKMARKS = 2;

    public static Single<History> fetchPost(final String boardName, final long threadId) {
        return watchThread(boardName, threadId).first(new History());
    }

    public static Flowable<History> watchThread(final String boardName, final long threadId) {
        From query = new Select()
                .from(History.class)
                .where(History.KEY_NAME + "=?", boardName)
                .where(History.KEY_THREAD_ID + "=?", threadId);

        Log.d(LOG_TAG, "SQL=" + query.toSql());
        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        return db.createQuery(History.TABLE_NAME, query.toSql(), (Object[]) query.getArguments())
                .toFlowable(BackpressureStrategy.BUFFER)
                .map(runQuery())
                .flatMap(History.flowableMapper())
                .flatMap((Function<List<History>, Flowable<History>>) histories -> {
                    if (histories == null || histories.size() == 0) {
                        return Flowable.just(new History());
                    }

                    return Flowable.just(histories.get(0));
                })
                .onErrorResumeNext((Function<Throwable, Flowable<History>>) throwable -> {
                    Log.e(LOG_TAG, "Error fetching history", throwable);
                    return Flowable.just(new History());
                });
    }

    public static Single<List<History>> fetchHistory() {
        return fetchHistory(null);
    }

    public static Single<List<History>> fetchHistory(final Boolean watched) {
        return fetchHistory(watched, 0);
    }

    public static Single<List<History>> fetchHistory(final Boolean watched, final int count) {
        return fetchHistory(watched, count, null);
    }

    public static Single<List<History>> fetchHistory(final Boolean watched, final int count, final Boolean showRemoved) {
        From query = new Select()
                .from(History.class)
                .orderBy(History.KEY_ORDER_ID + " ASC");

        if (showRemoved != null) {
            query = query.where(History.KEY_THREAD_REMOVED + "=?", showRemoved);
        }
        if (watched != null) {
            query = query.where(History.KEY_WATCHED + "=?", watched);
        }
        if (count > 0) {
            query = query.limit(count);
        }

        Log.d(LOG_TAG, "SQL=" + query.toSql());
        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        return db.createQuery(History.TABLE_NAME, query.toSql(), query.getArguments())
                .firstOrError()
                .map(runQuery())
                .flatMap(History.mapper())
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Error fetching history: watched=" + watched, throwable);

                    return Collections.emptyList();
                })
                .compose(DatabaseUtils.applySingleSchedulers());
    }

    public static Flowable<Boolean> setHistoryRemovedStatus(final String boardName, final long threadId, final boolean removed) {
        return Flowable.defer((Callable<Flowable<Boolean>>) () -> {

            BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
            BriteDatabase.Transaction transaction = db.newTransaction();

            int val = 0;

            try {
                ContentValues values = new ContentValues();
                values.put(History.KEY_THREAD_REMOVED, removed);

                val = db.update(History.TABLE_NAME, SQLiteDatabase.CONFLICT_REPLACE, values, History.KEY_THREAD_ID + "=?", String.valueOf(threadId));
                transaction.markSuccessful();
            } catch (Exception e) {
                Log.e(LOG_TAG, "Error deleting history: name=" + boardName + ", thread=" + threadId, e);
            } finally {
                transaction.end();
            }

            return Flowable.just(val > 0)
                    .onErrorReturn(throwable -> {
                        Log.e(LOG_TAG, "Error deleting history: name=" + boardName + ", thread=" + threadId, throwable);
                        return false;
                    })
                    .compose(DatabaseUtils.<Boolean>applySchedulers());
        });
    }

    public static Single<Boolean> putHistory(final String boardName, final ChanPost firstPost, final Integer postCount, final int lastReadPosition, final boolean isWatched, final int unreadCount) {
        if (firstPost == null) {
            return Single.just(false);
        }

        Log.d(LOG_TAG, "putHistory: name=" + boardName + ", thread=" + firstPost.getNo() + ", current watched=" + firstPost.isWatched() + ", new watched=" + isWatched);

        return Single.defer((Callable<Single<Boolean>>) () -> {
            BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
            BriteDatabase.Transaction transaction = db.newTransaction();
            long val = -1;

            try {
                final History history = new History();
                history.boardName = boardName;
                history.threadId = firstPost.getNo();
                history.userName = firstPost.getName();
                history.tim = firstPost.getTim();
                history.threadSize = postCount;
                history.lastReadPosition = lastReadPosition;
                history.watched = (byte) (isWatched ? 1 : 0);
                history.orderId = 0;
                history.unreadCount = unreadCount;

                final String text;
                if (!TextUtils.isEmpty(firstPost.getSubject())) {
                    text = firstPost.getSubject().toString();
                } else if (!TextUtils.isEmpty(firstPost.getCom())) {
                    text = firstPost.getCom();
                } else {
                    text = null;
                }
                history.text = text;
                history.lastAccess = System.currentTimeMillis();

                val = db.update(History.TABLE_NAME, SQLiteDatabase.CONFLICT_REPLACE, history.toContentValues(), history.clause(), history.vals());
                if (val <= 0) {
                    val = db.insert(History.TABLE_NAME, SQLiteDatabase.CONFLICT_REPLACE, history.toContentValues());
                    Log.d(LOG_TAG, "Added history: name=" + boardName + ", thread=" + firstPost.getNo() + ", watched=" + history.watched);
                } else {
                    Log.d(LOG_TAG, "Updated history: name=" + boardName + ", thread=" + firstPost.getNo() + ", watched=" + history.watched);
                }

                transaction.markSuccessful();

            } catch (Exception e) {
                Log.e(LOG_TAG, "Error adding history: name=" + boardName + ", thread=" + firstPost.getNo(), e);
            } finally {
                transaction.end();
            }

            return Single.just(val > 0);
        })
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Error adding history: name=" + boardName + ", thread=" + firstPost.getNo(), throwable);
                    return false;
                });
    }

    public static Single<Boolean> removeHistory(final String boardName, final long threadId) {

        if (boardName == null || threadId <= 0) {
            return Single.just(false);
        }

        From query = new Delete().from(History.class).where(History.KEY_THREAD_ID + "=?", threadId);
        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        return db.createQuery(History.TABLE_NAME, query.toSql(), query.getArguments())
                .firstOrError()
                .map(runQuery())
                .flatMap(History.mapper())
                .flatMap(validateHistoryDeleted())
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Error deleting history: name=" + boardName + ", thread=" + threadId, throwable);
                    return false;
                });
    }

    public static Single<Boolean> removeAllHistory(final boolean watched) {
        From query = new Delete().from(History.class).where(History.KEY_WATCHED + "=?", watched);
        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        return db.createQuery(History.TABLE_NAME, query.toSql(), query.getArguments())
                .firstOrError()
                .map(runQuery())
                .flatMap(History.mapper())
                .flatMap(validateHistoryDeleted())
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Error deleting all history", throwable);
                    return false;
                });
    }

    public static Single<Boolean> updateHistoryOrder(final List<History> historyList) {
        if (historyList == null || historyList.size() == 0) {
            Log.e(LOG_TAG, "Cannot update history order: history list is blank");
            return Single.just(false);
        }

        return Single.defer((Callable<Single<List<History>>>) () -> Single.just(historyList))
                .toFlowable()
                .flatMapIterable((Function<List<History>, Iterable<History>>) histories -> {
                    for (int i = 0; i < histories.size(); i++) {
                        histories.get(i).orderId = i;
                    }
                    return histories;
                }).single(new History())
                .flatMap((Function<History, Single<Boolean>>) history -> updateHistory(history))
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Error updating history order: size=" + historyList.size(), throwable);
                    return false;
                });
    }

    public static Single<Boolean> updateHistory(final History history) {
        if (history.threadId > -1) {
            return DatabaseUtils.updateTable(history);
        } else {
            return Single.just(false);
        }
    }

    public static Single<Boolean> pruneHistory(final int days) {
        final Long oldestHistoryTime = System.currentTimeMillis() - TimeUnit.DAYS.toMillis(days);
        From query = new Delete()
                .from(History.class)
                .where(History.KEY_LAST_ACCESS + "<?", oldestHistoryTime)
                .and(History.KEY_WATCHED + "=?", false);

        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        return db.createQuery(History.TABLE_NAME, query.toSql(), query.getArguments())
                .firstOrError()
                .map(runQuery())
                .flatMap(History.mapper())
                .flatMap(validateHistoryDeleted())
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Error deleting all history", throwable);
                    return false;
                });
    }

    public static Single<List<History>> getHistoryToPrune(final int days) {
        final Long oldestHistoryTime = System.currentTimeMillis() - TimeUnit.DAYS.toMillis(days);
        From query = new Select()
                .from(History.class)
                .where(History.KEY_LAST_ACCESS + "<?", oldestHistoryTime)
                .and(History.KEY_WATCHED + "=?", false);

        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        return db.createQuery(History.TABLE_NAME, query.toSql(), query.getArguments())
                .firstOrError()
                .map(runQuery())
                .flatMap(History.mapper())
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Error deleting all history", throwable);
                    return Collections.emptyList();
                });
    }

    public static Single<Boolean> setLastReadPost(final long threadId, final int pos) {
        From query = new Select().from(History.class).where(History.KEY_THREAD_ID + "=?", threadId);
        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        return db.createQuery(History.TABLE_NAME, query.toSql(), query.getArguments())
                .firstOrError()
                .map(runQuery())
                .flatMap(History.mapper())
                .toFlowable()
                .flatMapIterable((Function<List<History>, Iterable<History>>) histories -> histories)
                .flatMap((Function<History, Flowable<Boolean>>) history -> {
                    history.lastReadPosition = pos;
                    return DatabaseUtils.insertOrUpdateRow(history);
                })
                .single(false)
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Failed to set the last read position", throwable);
                    return false;
                })
                .compose(DatabaseUtils.applySingleSchedulers());

    }

    public static Single<Boolean> setUnreadCount(final long threadId, final int unread) {
        From query = new Select().from(History.class).where(History.KEY_THREAD_ID + "=?", threadId);
        BriteDatabase db = MimiApplication.getInstance().getBriteDatabase();
        return db.createQuery(History.TABLE_NAME, query.toSql(), query.getArguments())
                .firstOrError()
                .map(runQuery())
                .flatMap(History.mapper())
                .toFlowable()
                .flatMapIterable((Function<List<History>, Iterable<History>>) histories -> histories)
                .flatMap((Function<History, Flowable<Boolean>>) history -> {
                    history.unreadCount = unread;
                    return DatabaseUtils.insertOrUpdateRow(history);
                })
                .single(false)
                .onErrorReturn(throwable -> {
                    Log.e(LOG_TAG, "Failed to set the unread count", throwable);
                    return false;
                })
                .compose(DatabaseUtils.applySingleSchedulers());

    }


    private static Function<List<History>, Single<Boolean>> validateHistoryDeleted() {
        return histories -> Single.just(histories != null && histories.size() == 0);
    }
}

File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/HistoryTableConnection.kt
package com.emogoth.android.phone.mimi.db

import android.text.TextUtils
import android.util.Log
import com.emogoth.android.phone.mimi.db.ArchivedPostTableConnection.removeAllThreads
import com.emogoth.android.phone.mimi.db.ArchivedPostTableConnection.removeThreads
import com.emogoth.android.phone.mimi.db.MimiDatabase.Companion.getInstance
import com.emogoth.android.phone.mimi.db.models.History
import com.mimireader.chanlib.models.ChanPost
import io.reactivex.Flowable
import io.reactivex.Single
import io.reactivex.functions.BiFunction
import io.reactivex.functions.Function
import java.util.*
import java.util.concurrent.TimeUnit

object HistoryTableConnection {
    val LOG_TAG: String = HistoryTableConnection::class.java.simpleName
    const val HISTORY = 1
    const val BOOKMARKS = 2

    @JvmStatic
    fun fetchPost(boardName: String, threadId: Long): Single<History> {
        return observeThread(boardName, threadId).first(History())
    }

    @JvmStatic
    fun observeThread(boardName: String, threadId: Long): Flowable<History> {
        return getInstance()?.history()?.getHistoryByThread(boardName, threadId)
                ?: Flowable.just(History())
    }

    @JvmStatic
    fun fetchHistory(): Single<List<History>> {
        return observeHistory().first(emptyList()) ?: Single.just(emptyList())
    }

    @JvmStatic
    fun observeHistory(): Flowable<List<History>> {
        return getInstance()?.history()?.getAll() ?: Flowable.just(emptyList())
    }

    @JvmStatic
    fun fetchHistory(watched: Boolean): Single<List<History>> {
        return if (watched) {
            getInstance()?.history()?.getBookmarks()?.first(emptyList()) ?: Single.just(emptyList())
        } else {
            getInstance()?.history()?.getHistory()?.first(emptyList()) ?: Single.just(emptyList())
        }
    }

    @JvmStatic
    fun observeHistory(watched: Boolean): Flowable<List<History>> {
        return if (watched) {
            getInstance()?.history()?.getBookmarks() ?: Flowable.just(emptyList())
        } else {
            getInstance()?.history()?.getHistory() ?: Flowable.just(emptyList())
        }
    }

    @JvmStatic
    fun observeHistory(watched: Boolean, count: Int): Flowable<List<History>> {
        return if (watched) {
            getInstance()?.history()?.getBookmarks(count)
                    ?: Flowable.just(emptyList())
        } else {
            getInstance()?.history()?.getHistory(count)
                    ?: Flowable.just(emptyList())
        }
    }

    @JvmStatic
    fun fetchActiveBookmarks(count: Int): Single<List<History>> {
        return if (count > 0) {
            getInstance()?.history()?.getActiveBookmarks(count)?.first(emptyList())
                    ?: Single.just(emptyList())
        } else {
            getInstance()?.history()?.getActiveBookmarks()?.first(emptyList())
                    ?: Single.just(emptyList())
        }
    }

    @JvmStatic
    fun watchActiveBookmarks(count: Int): Flowable<List<History>> {
        return if (count > 0) {
            getInstance()?.history()?.getActiveBookmarks(count) ?: Flowable.just(emptyList())
        } else {
            getInstance()?.history()?.getActiveBookmarks() ?: Flowable.just(emptyList())
        }
    }

    private fun fetchHistorySortedByLastAccess(): Single<List<History>> {
        return getInstance()?.history()?.getAllByLastAccess()?.first(emptyList())
                ?: Single.just(emptyList())
    }

//    @JvmStatic
//    fun setThreadActive(boardName: String, threadId: Long, active: Boolean): Single<Boolean> {
//        return getInstance()?.history()?.threadActive(boardName, threadId, active)?.map { value: Int -> value > 0 }
//                ?: Single.just(false)
//    }

    @JvmStatic
    fun setHistoryRemovedStatus(boardName: String, threadId: Long, removed: Boolean): Single<Boolean> {
        return getInstance()?.history()?.threadRemoved(boardName, threadId, removed)?.map { value: Int -> value > 0 }
                ?: Single.just(false)
    }

    @JvmStatic
    fun setBookmark(boardName: String, threadId: Long, watched: Boolean): Single<Boolean> {
        return getInstance()?.history()?.watched(boardName, threadId, watched)?.map { value: Int -> value > 0 }
                ?: Single.just(false)
    }

    @JvmStatic
    fun setThreadSize(boardName: String, threadId: Long, size: Int): Single<Boolean> {
        return getInstance()?.history()?.threadSize(boardName, threadId, size)?.map { value: Int -> value > 0 }
                ?: Single.just(false)
    }

    @JvmStatic
    fun keepLatest(numberOfThreadsToKeep: Int): Single<Boolean> {
        return fetchHistorySortedByLastAccess()
                .flatMap { histories: List<History> ->
                    if (histories.size > numberOfThreadsToKeep) {
                        val prunedHistories: MutableList<History> = ArrayList(histories.size - numberOfThreadsToKeep)
                        var i = numberOfThreadsToKeep
                        while (i < histories.size) {
                            prunedHistories.add(histories[i])
                            i++
                        }
                        return@flatMap Single.just<List<History>>(prunedHistories)
                    } else {
                        return@flatMap Single.just<List<History>>(emptyList())
                    }
                }
                .flatMap { histories: List<History> ->
                    val threads: MutableList<Long> = ArrayList(histories.size)
                    for (history in histories) {
                        threads.add(history.threadId)
                    }
                    Single.just(threads)
                }
                .flatMap { threadList: List<Long> ->
                    if (threadList.isEmpty()) {
                        return@flatMap Single.just(false)
                    }
                    Single.zip(PostTableConnection.removeThreads(threadList), removeThreads(threadList),
                            BiFunction { aBoolean: Boolean, aBoolean2: Boolean -> aBoolean && aBoolean2 })
                }
    }

    @JvmStatic
    fun putHistory(boardName: String, threadId: Long, firstPost: ChanPost, postCount: Int): Single<Boolean> {
        return Single.defer {
            if (firstPost.no <= 0 || firstPost.no != threadId) {
                return@defer Single.just(false)
            }
            Log.d(LOG_TAG, "putHistory: name=" + boardName + ", thread=" + firstPost.no + ", current watched=" + firstPost.isWatched)
            val history = History()
            history.boardName = boardName
            history.threadId = threadId
            history.userName = firstPost.name ?: ""
            history.tim = firstPost.tim ?: ""
            history.threadSize = postCount
            history.lastReadPosition = 0
            history.watched = false
            history.orderId = 0
            history.unreadCount = 0
            val text: String = if (!TextUtils.isEmpty(firstPost.subject)) {
                firstPost.subject.toString()
            } else if (!TextUtils.isEmpty(firstPost.com)) {
                firstPost.com.toString()
            } else {
                ""
            }
            history.text = text
            history.lastAccess = System.currentTimeMillis()
            getInstance()?.history()?.insert(history)
            Single.just(true)
        }
    }

    @JvmStatic
    fun removeHistory(boardName: String, threadId: Long): Single<Boolean> {
        return Single.defer {
            getInstance()?.history()?.removeThread(boardName, threadId)
            Single.just(true)
        }
    }

    @JvmStatic
    fun removeAllHistory(watched: Boolean): Single<Boolean> {
        return Single.defer {
            getInstance()?.history()?.clear() ?: Single.just(false)
            Single.just(true)
        }
                .flatMap { PostTableConnection.removeAllThreads() }
                .flatMap { removeAllThreads() }
                .onErrorReturn { throwable: Throwable ->
                    Log.e(LOG_TAG, "Error deleting all history", throwable)
                    false
                }
    }

    @JvmStatic
    fun updateHistoryOrder(historyList: List<History>): Single<Boolean> {
        if (historyList.isEmpty()) {
            Log.e(LOG_TAG, "Cannot update history order: history list is blank")
            return Single.just(false)
        }
        return Flowable.defer { Flowable.just(historyList) }
                .flatMapIterable { histories: List<History> ->
                    for (i in histories.indices) {
                        histories[i].orderId = i
                    }
                    histories
                }
                .flatMap(Function { history: History -> setOrder(history.boardName, history.threadId, history.orderId).toFlowable() } as Function<History, Flowable<Boolean>>)
                .onErrorReturn { throwable: Throwable? ->
                    Log.e(LOG_TAG, "Error updating history order: size=" + historyList.size, throwable)
                    false
                }
                .toList()
                .map { booleans: List<Boolean> ->
                    var success = true
                    for (b in booleans) {
                        if (success) {
                            success = b
                        }
                    }
                    success
                }
    }

    @JvmStatic
    fun updateHistory(history: History): Single<Boolean> {
        return if (history.threadId == 0L) {
            Single.just(false)
        } else Single.defer { Single.just(getInstance()?.history()?.update(history) ?: 0 > 0) }
    }

    @JvmStatic
    fun setOrder(boardName: String, threadId: Long, order: Int): Single<Boolean> {
        return getInstance()?.history()?.order(boardName, threadId, order)?.map { value: Int -> value > 0 }
                ?: Single.just(false)
    }

    @JvmStatic
    fun pruneHistory(days: Int): Single<Boolean> {
        val oldestHistoryTime = System.currentTimeMillis() - TimeUnit.DAYS.toMillis(days.toLong())
        return Single.defer {
            getInstance()?.history()?.prune(oldestHistoryTime, false)
                    ?: return@defer Single.just(false)
            Single.just(true)
        }
    }

    @JvmStatic
    fun getHistoryToPrune(days: Int): Single<List<History>> {
        val oldestHistoryTime = System.currentTimeMillis() - TimeUnit.DAYS.toMillis(days.toLong())
        return getInstance()?.history()?.getOldHistory(oldestHistoryTime, false)
                ?.onErrorReturn { throwable: Throwable? ->
                    Log.e(LOG_TAG, "Error deleting history from $days day(s) ago", throwable)
                    emptyList()
                } ?: Single.just(emptyList())
    }

    @JvmStatic
    fun setLastReadPos(boardName: String, threadId: Long, pos: Int): Single<Boolean> {
        return getInstance()?.history()?.lastReadPos(boardName, threadId, pos)?.map { value: Int -> value > 0 }
                ?.onErrorReturn { throwable: Throwable? ->
                    Log.e(LOG_TAG, "Failed to set the last read position to $pos", throwable)
                    false
                } ?: Single.just(false)
    }

    @JvmStatic
    fun setUnreadCount(boardName: String, threadId: Long, unread: Int): Single<Boolean> {
        return getInstance()?.history()?.unreadCount(boardName, threadId, unread)?.map { value: Int -> value > 0 }
                ?.onErrorReturn { throwable: Throwable? ->
                    Log.e(LOG_TAG, "Failed to set the unread count to $unread", throwable)
                    false
                } ?: Single.just(false)
    }

//    @JvmStatic
//    fun nextRefresh(bookmarksOnly: Boolean): Single<History> {
//        return if (bookmarksOnly) {
//            getInstance()?.history()?.oldestActiveBookmark() ?: Single.just(History())
//        } else {
//            getInstance()?.history()?.oldestActiveHistory() ?: Single.just(History())
//        }
////        return getInstance()?.history()?.oldestActiveHistory() ?: Single.just(History())
//    }

    @JvmStatic
    private fun validateHistoryDeleted(): Function<List<History>, Single<Boolean>> {
        return Function { histories: List<History> -> Single.just(histories.isEmpty()) }
    }
}